akka-remote/src/main/scala/akka/remote/artery/Codecs.scala

```diff
@@ -156,35 +137,32 @@ private[remote] class Decoder(
         val envelope = grab(in)
         envelope.parseHeader(headerBuilder)
 
-        // FIXME: Instead of using Strings, the headerBuilder should automatically return cached ActorRef instances
-        // in case of compression is enabled
-        // FIXME: Is localAddress really needed?
-
-        val sender =
-          if (headerBuilder.isNoSender)
-            OptionVal.None
-          else {
-            senderCache.get(headerBuilder.senderActorRef) match {
-              case null ⇒
-                val ref = resolveActorRefWithLocalAddress(headerBuilder.senderActorRef)
-                // FIXME this cache will be replaced by compression table
-                if (senderCache.size() >= 1000)
-                  senderCache.clear()
-                senderCache.put(headerBuilder.senderActorRef, ref)
-                OptionVal(ref)
-              case ref ⇒ OptionVal(ref)
-            }
-          }
-
-        val recipient =
-          if (headerBuilder.isNoRecipient)
-            OptionVal.None
-          else
-            resolveRecipient(headerBuilder.recipientActorRef)
-
         val originUid = headerBuilder.uid
         val association = inboundContext.association(originUid)
 
+        val recipient: OptionVal[InternalActorRef] = headerBuilder.recipientActorRef match {
+          case OptionVal.Some(ref) ⇒ OptionVal(ref.asInstanceOf[InternalActorRef])
+          case _                   ⇒ resolveRecipient(headerBuilder.recipientActorRefPath)
+        }
+
+        val sender: InternalActorRef = headerBuilder.senderActorRef match {
+          case OptionVal.Some(ref) ⇒ ref.asInstanceOf[InternalActorRef]
+          case _                   ⇒ resolveActorRefWithLocalAddress(headerBuilder.senderActorRefPath)
```

Comment:
in the `case _ =>` the value of `senderActorRef` is `OptionVal.None`, i.e. it was not compressed.
Compression acts like a cache here, in the sense of that we directly get the actual ActorRef back.
If we don't, we need to get the "raw string" and resolve it.

Not sure if we'd rather mvoe this into the header builder... it would need more context AFAICS (what local address is)